src/librbd/api/Image.cc

```diff
@@ -168,33 +169,40 @@ int Image<I>::get_parent(I *ictx,
 
   RWLock::RLocker snap_locker(ictx->snap_lock);
   RWLock::RLocker parent_locker(ictx->parent_lock);
-  if (ictx->parent == nullptr) {
-    return -ENOENT;
-  }
 
-  cls::rbd::ParentImageSpec parent_spec;
-  if (ictx->snap_id == CEPH_NOSNAP) {
-    parent_spec = ictx->parent_md.spec;
-  } else {
-    r = ictx->get_parent_spec(ictx->snap_id, &parent_spec);
-    if (r < 0) {
-      lderr(cct) << "error looking up snapshot id " << ictx->snap_id << dendl;
-      return r;
+  auto parent = ictx->parent;
+  bool release_parent_locks = false;
+  BOOST_SCOPE_EXIT_ALL(ictx, &parent, &release_parent_locks) {
+    if (release_parent_locks) {
+      ictx->parent->parent_lock.put_read();
+      ictx->parent->snap_lock.put_read();
     }
-    if (parent_spec.pool_id == -1) {
-      return -ENOENT;
+  };
+
+  // if a migration is in-progress, the true parent is the parent
+  // of the migration source image
+  if (!ictx->migration_info.empty() && parent != nullptr) {
+    if (parent != nullptr) {
```

Comment:
nit: no need to check  for `nullptr` here as it is already checked one line above.